src/sql/parts/notebook/notebookInput.ts

```diff
@@ -8,150 +8,189 @@
+		@INotebookService private notebookService: INotebookService
+	) {
 		super();
-		this.dirty = false;
-		this._providerId = provider;
-		this._standardKernels = [];
-	}
-
-	public get providerId(): string {
-		return this._providerId;
-	}
-
-	public set providerId(value: string) {
-		this._providerId = value;
-	}
-
-	public get providers(): string[] {
-		return this._providers;
-	}
-
-	public set providers(value: string[]) {
-		this._providers = value;
-	}
-
-	public get connectionProfileId(): string {
-		return this._connectionProfileId;
-	}
-
-	public get standardKernels(): IStandardKernelWithProvider[] {
-		return this._standardKernels;
-	}
-
-	public set standardKernels(value: IStandardKernelWithProvider[]) {
-		value.forEach(kernel => {
-			this._standardKernels.push({
-				connectionProviderIds: kernel.connectionProviderIds,
-				name: kernel.name,
-				notebookProvider: kernel.notebookProvider
-			});
-		});
-	}
-
-	public get defaultKernel(): azdata.nb.IKernelSpec {
-		return this._defaultKernel;
-	}
+		this._register(this.notebookService.onNotebookEditorAdd(notebook => {
+			if (notebook.id === this.notebookUri.toString()) {
+				// Hook to content change events
+				notebook.modelReady.then(() => {
+					this._register(notebook.model.contentChanged(e => this.updateModel()));
+					this._register(notebook.model.kernelChanged(e => this.updateModel()));
+				}, err => undefined);
+			}
+		}));
 
-	public set defaultKernel(kernel: azdata.nb.IKernelSpec) {
-		this._defaultKernel = kernel;
+		if (this.textEditorModel instanceof UntitledEditorModel) {
+			this._register(this.textEditorModel.onDidChangeDirty(e => this.setDirty(this.textEditorModel.isDirty())));
+		} else {
+			this._register(this.textEditorModel.onDidStateChange(e => this.setDirty(this.textEditorModel.isDirty())));
+		}
+		this.dirty = this.textEditorModel.isDirty();
 	}
 
-	get isTrusted(): boolean {
-		return this._isTrusted;
-	}
-
-	get onDidChangeDirty(): Event<void> {
-		return this._onDidChangeDirty.event;
+	public get contentString(): string {
+		let model = this.textEditorModel.textEditorModel;
+		return model.getValue();
 	}
 
 	get isDirty(): boolean {
-		return this.dirty;
+		return this.textEditorModel.isDirty();
 	}
 
 	public setDirty(dirty: boolean): void {
 		if (this.dirty === dirty) {
 			return;
 		}
-
 		this.dirty = dirty;
 		this._onDidChangeDirty.fire();
 	}
 
-	save(): TPromise<boolean> {
-		if (this.saveHandler) {
-			return TPromise.wrap(this.saveHandler(this.handle));
+	public updateModel(): void {
+		let notebookModel = this.getNotebookModel();
+		if (notebookModel && this.textEditorModel && this.textEditorModel.textEditorModel) {
+			let content = JSON.stringify(notebookModel.toJSON(), undefined, '    ');
+			let model = this.textEditorModel.textEditorModel;
+			let endLine = model.getLineCount();
+			let endCol = model.getLineLength(endLine);
+			this.textEditorModel.textEditorModel.applyEdits([{
+				range: new Range(1, 1, endLine, endCol),
```

Comment:
How does this work? Does this say that the entirety is edited every time, or is the textEditorModel smart enough to diff what the changes are?

Reply:
TextEditorModel handles the changes with applyEdits. That's precisely the reason we are providing start, end and offset positions.

Reply:
To clarify - yes the entirety is edited every time, but if you look at the API it says it's not added to the undo stack in this case. It's necessary to do this in order to ensure the dirty state is updated correctly in the document.